### PR TITLE
fix(compiler): detect and strip data- prefix

### LINKED ISSUE #2687: Issue with data-* properties not working on elements.
I have been playing around with Angular2  (and enjoying it) a fair bit recently but have noticed what I think is perhaps a bug but not quite sure.

The problem is to do with using custom data-* properties as I am never able to get them to properly display. 

For instance the following code example works when replacing [title] with either [lang], [alt] or [id] then I will still get a rendered output on my element. However whenever I put any custom element replacing [title] say with [data-list] then no string interpolation happens.

I was wondering if this is because data-* attributes aren't specified in the _HTMLElementPropertyList?

Would appreciate any guidance as I have repeatedly read over the following documentation to no avail:

https://github.com/angular/angular/blob/master/modules/angular2/docs/core/01_templates.md

## index.html

```html
<!DOCTYPE html>
<html>

  <head>
    <script src="https://jspm.io/system@0.16.js"></script>
    <script src="https://code.angularjs.org/2.0.0-alpha.27/angular2.dev.js"></script>
  </head>

  <body>
    <app>...</app>
    <script>
      System.config({
        paths: {
          '*': '*.js',
          'angular2/*': 'angular2/*',
        }
      });
      System.import('main');
    </script>
  </body>

</html>
```

## bootstrap.js

```ts
import {Component, View, bootstrap, Attribute} from 'angular2/angular2';

@Component({
  selector: 'table',
  properties: {'name':'name'}
})
@View({
  template:`<table [title]="_name">Hello {{_name}}</table>`
})
class Hello {
  _name: string;
  constructor(@Attribute('name') value:string) {
    this._name = value;
  };
}

@Component({
  selector: 'app',
})
@View({
  template:
  `
    <div>
      <table name="testing"></table>
    </div>
  `,
  directives: [Hello]
})
class Application {
  constructor() { };
}

bootstrap(Application);
```

### PR DESCRIPTION
**Pull request by pkozlowski-opensource**
Fixes #2687

[Review on Reviewable](https://reviewable.io/reviews/angular/angular/2719)

### CODE REVIEW

File: modules/angular2/test/render/dom/compiler/property_binding_parser_spec.ts [hunk 1]
```diff
@@ -34,6 +34,11 @@ export function main() {
       expect(results[0].propertyBindings.get('a').source).toEqual('b');
     });
 
+    it('should detect [] syntax with data- prefix', () => {
+      var results = process(el('<div data-[a]="b"></div>'));
```
Comment (vicb) [inline modules/angular2/test/render/dom/compiler/property_binding_parser_spec.ts]: I'm unsure we want this for `[], () and [()]` ?

**Comment by vicb** [comment]
As far I understand the original issue, the pb is that `[data-foo]="bar"`.

I don't think this PR solve the issue, does it ?

_Note: I don't think `(data-foo)=...` or `[(data-foo)]=...` make any sense, maybe we should just throw ?_

**Comment by pkozlowski-opensource** [comment]
right, the original issue was for `[data-foo]` but this would imply binding to `dataFoo` property - which doesn't make sense. IMO prefixing with `data-` is there to make HTML validators happy and this PR allows just this. We might throw for other use-cases just to warn users, to be seen.

**Comment by vicb** [comment]
> but this would imply binding to dataFoo property 

Shouldn't we special case this to bind to `dataset.foo` ?

**Comment by SebastianM** [comment]
As @pkozlowski-opensource said:

> IMO prefixing with data- is there to make HTML validators happy this PR allows just this.

I'm not sure if this fixes the valdiation problem. When validating this example with the [W3C Validator](https://validator.w3.org/nu/):

```html
<!DOCTYPE html>
<html>
  <head>
    <title>demo</title>
  </head>
  <body>
    <div data-[abc]="abc"></div>
  </body>
</html>
```
I get this Error :
html5

The HTML5 Spec for data-* attributes:
> A custom data attribute is an attribute in no namespace whose name starts with the string "data-", has at least one character after the hyphen, is **XML-compatible**, and contains no uppercase ASCII letters.

[Source](http://www.w3.org/TR/html5/dom.html#embedding-custom-non-visible-data-with-the-data-*-attributes)

**Comment by pkozlowski-opensource** [comment]
@SebastianM what you are saying is true, but:
* proposed fix will work for other forms of bindings, ex.: `data-bind-foo="exp"`
* as far as I recall there are some inconsistencies in specs, those "special" chars should be allowed in the end

@vicb I'm not sure we should have any special casing for `[data-foo]="exp"` - IMO this form doesn't bring anything for HTML validators and people shouldn't use it. IMO we should throw in this case and we do so currently.

**Comment by mhevery** [comment]
Agree with @pkozlowski-opensource that we should only strip out `data-[foo]` case.
@SebastianM if you want to make validators happy you could write `data-bind-foo`

**Comment by mhevery** [comment]
Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

<sup>**[modules/angular2/test/render/dom/compiler/property_binding_parser_spec.ts, line 38 \[r1\]](https://reviewable.io:443/reviews/angular/angular/2719#-JsgWX5RFocVeCQ7RpkE-r1-38)** ([raw file](https://github.com/angular/angular/blob/50f663dabee7706fc2b3173d2ff51b115ee8c386/modules/angular2/test/render/dom/compiler/property_binding_parser_spec.ts#L38)):</sup>
I think the test is correct.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/angular/angular/2719)

**Comment by vicb** [comment]
I don't agree and I've removed the merge flag.

I can't understand why `data-[foo]` should bind to foo in angular.

`data-foo` do not set the `foo`attribute in HTML.

Needs more discussion before merging.

**Comment by gkalpak** [comment]
Other than (outdated) validators, is there any other reason for stripping `data-` ?
Is it such a strong use-case ?
Won't validators be broken anyway (from other things such as custom tags) ?

**Comment by pkozlowski-opensource** [comment]
@vicb would you suggest removing it only before `bind-` and `on-`? What are you proposing? 

For me the system is simple - we strip `data-` prefix and then apply things as usual. Simple to understand and covering all the use-cases.

But I don't mind really - I don't use HTML validators myself. Seems like it is a big deal for some people, tough.

**Comment by vicb** [comment]
I suggested https://github.com/angular/angular/pull/2719#issuecomment-114895596 & https://github.com/angular/angular/pull/2719#discussion_r33150636

Out of angular context `<x data-foo="bar">` would set `x.dataset.foo` (& `x.getAttribute('data-foo') == 'bar'`). I think this PR deviates from the standard html behavior.

IMO the only thing we need to support (hence special case) is `[data-foo]` / `bind-data-foo`

**Comment by pkozlowski-opensource** [comment]
> IMO the only thing we need to support (hence special case) is [data-foo] / bind-data-foo

I'm sorry but I disagree here... `[data-foo]` and `bind-data-foo` imply "bind to `dataFoo`" property which is an error as such property doesn't exist. We should be throwing. I don't see any reason of special-casing this - this would make the whole system less consistent and would introduce one more special case to remember / explain.

This is just my opinion of course, don't have more arguments so I suggest that you and @mhevery reach an agreement here and proceed accordingly.

**Comment by tbosch** [comment]
@vicb @mhevery ping, what should be done here?

**Comment by vicb** [comment]
@pkozlowski-opensource let's have a chat to discuss about this

**Comment by vicb** [comment]
We have clarified this with @pkozlowski-opensource, LGTM.

There will be a new syntax `[data.foo]` to bind data attributes to expression.

**Comment by vicb** [comment]
@pkozlowski-opensource could you please create a separate issue for `[data.foo]`

**Comment by gdi2290** [comment]
In my opinion the `DomRenderer` should output `data-` as an W3C Validator option rather than compiler being aware of something that only matters for browsers

**Pull request closed without merging**